### DISCUSSION TITLE
Scheme and CLR Exception Handling

### DISCUSSION THREAD

**Discussion by bcalco** [leppie/IronScheme#108] [Discussions] [answered]
@leppie et al...

I'm working on a little unit testing harness, and am finding it hard to handle exceptions in a way that allows tests that throw them to do so without halting test script execution altogether. 

The following three forms create a handy little DSL for tests that works great until something throws an exception. The `(try ...)` form is intended to catch when an exception is raised in the `thunk`, which is a user-defined expression that is evaluated:

```scheme
  (define (try thunk)
    (display "Trying! ")
    (call/cc 
      (lambda (k)
        (with-exception-handler
          (lambda (ex)
            (if (error? ex) 
                (begin (display "Error!")
                       (newline) 
                       (k "Error!"))
                (begin (if (violation? ex)
                           (begin (display "Violation!")
                                  (newline)
                                  (k "Violation!"))
                           (begin (display "Something worse than violation!")
                                  (newline)
                                  (k "Something worse than violation!"))))))
          (lambda () thunk)))))

  (define exec-expr
    (lambda (expr equality-test expected context explanation)
      (let* ([actual expr]
              [success? (equality-test actual expected)])
        (update-test-results success? context expected actual explanation))))

  (define-syntax check
    (syntax-rules (that from is exactly numerically equal equivalent to because)

      [(check that expr from context is equal to expected because explanation)
       (exec-expr (try expr) equal? expected context explanation)]

      [(check that expr from context is exactly equal to expected because explanation)
       (exec-expr (try expr) eq? expected context explanation)]

      [(check that expr from context is equivalent to expected because explanation)
       (exec-expr (try expr) eqv? expected context explanation)]

      [(check that expr from context is numerically equal to expected because explanation)
       (exec-expr (try expr) = expected context explanation)]))
```

A contrived example of a call that raises an exception looks like this:

```scheme
(check that
    (raise-continuable 'kaboomerang)          
 from fun-defs
 is equal to "Exception: kaboom" ; this will fail cuz expect what (try ... ) returns ....
 because "We don't want exceptions halting execution of test scripts.")
```
It doesn't matter whether I call `(raise...)` or `(raise-continuable...)`; nor does it matter if I wrap the expression in `(try ...)`, the result is always the same (shown here in the context of one of my smaller test suits, the one testing SRFI 219 as it happens):

```
------------------------------------------
-- Testing alternate ways to define functions.
------------------------------------------

Trying! ((compose1 sqr sin) 2) should equal 0.826821810431806. => PASS
Trying! ((compose2 sqr sin) 2) should equal 0.826821810431806. => PASS
Trying! ((compose3 sqr sin) 2) should equal 0.826821810431806. => PASS
Trying! (((compose-add sqr sin) 1) 1)) should equal 0.826821810431806. => PASS
Trying! (((compose-add* sqr sin) 1) 0.5 0.5) should equal 0.826821810431806. => PASS
Trying! ((compose+ sqr sin) 0.5 0.5 0.5 0.5) should equal 0.826821810431806. => PASS
Trying! ((((compose-add-multi-canonical sqr sin) 0.4) 0.4) 0.4 0.4 0.4) should equal 0.826821810431806. => PASS
Trying! ((((compose-add-multi sqr sin) 0.4) 0.4) 0.4 0.4 0.4) should equal 0.826821810431806. => PASS
kaboomerang
```
If I remove the failing test, it completes like so:

```

------------------------------------------
-- Testing alternate ways to define functions.
------------------------------------------

Trying! ((compose1 sqr sin) 2) should equal 0.826821810431806. => PASS
Trying! ((compose2 sqr sin) 2) should equal 0.826821810431806. => PASS
Trying! ((compose3 sqr sin) 2) should equal 0.826821810431806. => PASS
Trying! (((compose-add sqr sin) 1) 1)) should equal 0.826821810431806. => PASS
Trying! (((compose-add* sqr sin) 1) 0.5 0.5) should equal 0.826821810431806. => PASS
Trying! ((compose+ sqr sin) 0.5 0.5 0.5 0.5) should equal 0.826821810431806. => PASS
Trying! ((((compose-add-multi-canonical sqr sin) 0.4) 0.4) 0.4 0.4 0.4) should equal 0.826821810431806. => PASS
Trying! ((((compose-add-multi sqr sin) 0.4) 0.4) 0.4 0.4 0.4) should equal 0.826821810431806. => PASS


------------------------------------------
-- Final Results
------------------------------------------

Passed: 8
Failed: 0

```

So it is doing what I don't want it to do - halting the entire test suite at the point of the raised exception. I would like to report a test failure, ideally with a message from the exception. I get that some kinds of exceptions are unrecoverable (like a syntax error) but short of that a unit test suite should not be halted by catchable exceptions so I'm kinda stuck here... and it is kinda urgent.

Any advice here would be appreciated!

**Comment by leppie** [comment]
I think I see the problem. Your `try` proc need to be a macro, else `expr` is evaluated before the `try` proc is called. 

Also `guard` would be better than your `try`. You can just wrap your `try` macro around that for much cleaner code. 

Also `raise` is safer as it requires you to handle the exception. `raise-continuable` is likely for something that will trigger the continuations limitation on IronScheme.

There are several test suites in the repo already for R6RS, SRFI's and other libraries that does the correct thing.

**Reply by bcalco** [reply]
I will refactor this as you suggest here but your suggestion below got it working. Fresh eyes make all the difference sometimes!

**Comment by leppie** [comment, marked as answer]
Alternatively, you can do this instead:

`(try (lambda () expr))` instead of `(try expr)` 

And then just `thunk` instead of `(lambda () thunk)` in your `try` proc. 

Generally thunk means `(lambda () expr)`.

**Reply by bcalco** [reply]
This got me unstuck immediately. Thanks for the catch!

![image](https://user-images.githubusercontent.com/14943929/140508220-133012b5-99b0-486b-a23b-8556fd45fda7.png)

**Comment by bcalco** [comment]
@leppie - any particular advice or examples of capturing CLR exceptions? Particularly pulling info out of them that can be displayed in relatively informative ways.

**Reply by leppie** [reply]
You should have searched the repo for `guard`, and you might have noticed `clr-guard` in the `(ironscheme clr)` library :D

**Comment by bcalco** [comment]
@leppie OK so I'm trying to create a macro that does what I want and am a little befuddled by the error I'm getting.

I'm importing `(wak trc-testing display-condition)` (which is REALLY slowing down execution at startup as there must be a lot of macro magic going on in that library), and trying to use the guard approach, as below:

```scheme
  (define-syntax try-guard
    (syntax-rules ()
      [(_ e1 e2 ...)
       (guard (x [(condition? x) 
                    (let ([p (open-output-string)])
                      (begin (display-condition x p)
                             (get-output-string p)))]) e1 e2 ...)]))

  (define exec-expr
    (lambda (expr equality-test expected context explanation)
      (let* ([actual expr]
              [success? (equality-test actual expected)])
        (update-test-results success? context expected actual explanation))))

  (define-syntax check
    (syntax-rules (that from is exactly numerically equal equivalent to because)

      [(check that expr from context is equal to expected because explanation)
       (exec-expr (try-guard expr) equal? expected context explanation)]

      [(check that expr from context is exactly equal to expected because explanation)
       (exec-expr (try-guard expr) eq? expected context explanation)]

      [(check that expr from context is equivalent to expected because explanation)
       (exec-expr (try-guard expr) eqv? expected context explanation)]

      [(check that expr from context is numerically equal to expected because explanation)
       (exec-expr (try-guard expr) = expected context explanation)]))
```

However I'm getting the following error:

```
&who: get-output-string
&message: "unbound identifier"
&undefined
&source-information:
  file-name: "Z:\\Documents\\GitHub\\Apex.Unify\\Build\\VisualScheme\\lib\\unit-testing.sls"
  character: "(83,31) - (83,48)"
&trace: #<syntax get-output-string ((83,31) - (83,48) of
 Z:\Documents\GitHub\Apex.Unify\Build\VisualScheme\lib\unit-testing.sls)>
&trace: #<syntax (try-guard ((compose1 sqr sin) 2))>
&trace: #<syntax (check that ((compose1 sqr sin) 2) from fun-defs is numerically equal to 0.826821810431806 because
 "((compose1 sqr sin) 2) should equal 0.826821810431806.") ((81,7) - (85,74) of
 Z:\Documents\GitHub\Apex.Unify\Build\Common\SchemeTests\lib\miscellaneous\function-definitions.sls)>
```

I use this approach to capture text output when employing `(pretty-print ...)` elsewhere so I'm not sure how it's possible that `get-output-string` is an unbound identifier. Any insights welcome!

**Reply by leppie** [reply]
> I'm importing `(wak trc-testing display-condition)` (which is REALLY slowing down execution at startup as there must be a lot of macro magic going on in that library)

The author of that code really loves macros. He also wrote `foof-loop`. Some crazy stuff, but I am not complaining as IronScheme handles it fine. TRC is pretty heavyweight though. 

> I use this approach to capture text output when employing `(pretty-print ...)` elsewhere so I'm not sure how it's possible that `get-output-string` is an unbound identifier. Any insights welcome!

Scheme conditions (aka exceptions) are compound, so they have parts. `display-condition` actually does the right thing. 

What you are looking for is `condition-message` (see the record pattern?). See http://www.r6rs.org/final/html/r6rs-lib/r6rs-lib-Z-H-8.html#node_idx_412 .

Scheme/R6RS conditions are pretty complex, but are quite easy to use once you understand them.

Now that I looked at this message, this is not a CLR exception (in case you possibly confused it with your previous question).  If you dont catch a CLR exception, it will bubble all the way to the top, and you will get it in code, or a nice red `Exception.ToString()` in the REPL. 

Example:

```
> (clr-call string Substring "hello" 42)
Unhandled CLR exception during evaluation:
CLR Exception: System.ArgumentOutOfRangeException
System.ArgumentOutOfRangeException: startIndex cannot be larger than length of string.
Parameter name: startIndex
   at System.String.Substring(Int32 startIndex, Int32 length)
   at eval-core(006).Initialize(CodeContext )
   at #.ironscheme.exceptions::dynamic-wind(Object in, Object proc, Object out)
   at #.ironscheme.exceptions::dynamic-wind(Object in, Object proc, Object out)
   at IronScheme.Runtime.Builtins.CallWithCurrentContinuation(Object fc1)
   at IronScheme.Runtime.R6RS.Exceptions.WithClrExceptionHandler(Object handler, Object thunk)
```

Those can be dealt with with `clr-guard`, but you need to obviously use CLR calls to get the details.

Example:

```
> (clr-guard (ex (#t (clr-prop-get Exception Message ex)))  (clr-call string Substring "hello" 42))
"startIndex cannot be larger than length of string.\r\nParameter name: startIndex"
```

Hope this helps!

**Reply by leppie** [reply]
I tried to cover all the builtin R6RS and IronScheme libs to raise Scheme conditions, but let me know if I missed any (some SRFI's might not deal with it correctly).

Examples:

```
> (substring "hello" 42)
Unhandled exception during evaluation:
&assertion
&who: ironscheme.base::substring
&message: "invalid argument count, expected 3 got 2"
&irritants: ("hello" 42)

> (substring "hello" 0 42)
Unhandled exception during evaluation:
&assertion
&who: substring
&message: "outside string bounds"
&irritants: (42)
```

But as soon as you use `clr-call` and friends, you need to cover exceptional cases yourself.

Unfortunately stuff like this does not work as expected:
```
> (clr-guard (ex (#t (condition-message ex)))  (clr-call string Substring "hello" 42))
#f
```

At least it is a lot more consistent than say Clojure where everything is just JVM exceptions. Granted, Clojure does not really have a specification they have to stick to.

**Reply by leppie** [reply]
Just some more examples:

```
> (guard (ex (#t (condition-message ex)))  (substring "hello" 0 42))
"outside string bounds"
```

Also no need to call `display-condition`, IronScheme already does that.
```
> (guard (ex (#t (display ex)))  (substring "hello" 0 42))
&assertion
&who: substring
&message: "outside string bounds"
&irritants: (42)
```

If you need to deal with syntax, it gets more tricky, you cant just use it as is:
```
> (guard (ex (#t (display ex)))  (define "hello" 42))
Unhandled exception during evaluation:
&who: define
&message: "invalid syntax"
&syntax:
  form: (define "hello" 42)
  subform: #f
&trace: #<syntax (define "hello" 42)>
```

To catch syntax exceptions you would need to wrap them in `eval` . The R6RS test suite does this.

```
> (guard (ex (#t (condition-message ex))) (eval '(define "hello" 42) (interaction-environment)))
"invalid syntax"
```

**Reply by leppie** [reply]
> ```scheme
> (let ([p (open-output-string)])
>                       (begin (display-condition x p)
>                              (get-output-string p)))])
> ```

Hint: `call-with-string-output-port`  https://github.com/IronScheme/IronScheme/blob/master/IronScheme/IronScheme.Console/build/io/ports.sls#L898

**Comment by leppie** [comment]
This has actually turned out to be a good primer on dealing with Scheme and CLR exceptions :D

**Reply by bcalco** [reply]
I will review your above feedback, thanks again for it! I am sure I'll have more questions.

My CLR exception question had nothing to do with the case I was complaining about, above. That was simply not understanding why get-string-output was being called out as an unaabound identifier in that context. Still don't understand that. My CLR concern has to do with some higher level stuff I'm about to embark on, if I can get this macro done and working reliably.

**Reply by leppie** [reply]
If you search the repo for `get-string-output` you will see it is not used or defined anywhere.

`get-output-string` is defined though :D

**Reply by leppie** [reply]
BTW the REPL has completion (based on all the imported definitions), so typing in `get-str` and pressing  TAB will present all the options (not quite as nice as Chez listing them all), but just keep on pressing TAB :D

**Reply by bcalco** [reply]
it was actually get-output-string that was getting the rror:

```
&who: get-output-string
&message: "unbound identifier"
&undefined
&source-information:
  file-name: "Z:\\Documents\\GitHub\\Apex.Unify\\Build\\VisualScheme\\lib\\unit-testing.sls"
  character: "(83,31) - (83,48)"
&trace: #<syntax get-output-string ((83,31) - (83,48) of
 Z:\Documents\GitHub\Apex.Unify\Build\VisualScheme\lib\unit-testing.sls)>
&trace: #<syntax (try-guard ((compose1 sqr sin) 2))>
&trace: #<syntax (check that ((compose1 sqr sin) 2) from fun-defs is numerically equal to 0.826821810431806 because
 "((compose1 sqr sin) 2) should equal 0.826821810431806.") ((82,7) - (86,74) of
 Z:\Documents\GitHub\Apex.Unify\Build\Common\SchemeTests\lib\miscellaneous\function-definitions.sls)>
```

My description had the typo, not my code (for a change).

**Comment by bcalco** [comment]
Also one point of clarification: I am not wanting just to dump to display but to capture a string so I can reuse later when listing the errors after the final output report.

**Reply by leppie** [reply]
I think I showed that in the examples above.

**Comment by bcalco** [comment]
Ok so I'm using (display-condition ...) to print out the error and it works as long as the condition is continuable. So `(raise-continuable 'kaboomerang)` prints and continues running but `(raise 'kaboomerang)` does not: it prints and halts with 
```
Z:\Documents\GitHub\Apex.Unify\Tests\sparql>adb run -s test-misc.scm


------------------------------------------
-- Testing alternate ways to define functions.
------------------------------------------

Trying! ((compose1 sqr sin) 2) should equal 0.826821810431806. => PASS
Trying! ((compose2 sqr sin) 2) should equal 0.826821810431806. => PASS
Trying! ((compose3 sqr sin) 2) should equal 0.826821810431806. => PASS
Trying! (((compose-add sqr sin) 1) 1)) should equal 0.826821810431806. => PASS
Trying! (((compose-add* sqr sin) 1) 0.5 0.5) should equal 0.826821810431806. => PASS
Trying! ((compose+ sqr sin) 0.5 0.5 0.5 0.5) should equal 0.826821810431806. => PASS
Trying! ((((compose-add-multi-canonical sqr sin) 0.4) 0.4) 0.4 0.4 0.4) should equal 0.826821810431806. => PASS
Trying! ((((compose-add-multi sqr sin) 0.4) 0.4) 0.4 0.4 0.4) should equal 0.826821810431806. => PASS
Trying! Uh oh... Non-condition object: kaboomerangkaboomerang
&non-continuable
&message: "handler returned"
```

I just want a generic handler that will catch anything, return as a string so it can be included, and never keels over no matter what's thrown at it.

**Reply by bcalco** [reply]
OK so I figured out what I was doing wrong there.  I will consolidate my learning from you responses and re-engage on the other side of some experiments. Thanks for all you help thus far, @leppie.

BTW, there is a CLR "not implemented" exception related to something re: bytevectors IIRC that I will need you to look at next but it seems like a proper issue which I thought I'd logged but apparently did not. So I'll get the details on that and log them in Issues.

**Reply by leppie** [reply]
> BTW, there is a CLR "not implemented" exception related to something re: bytevectors IIRC

Could be related to new overloads in newer .NET versions, In that case I would just need to annotate the `clr-call` to choose the correct one. I recall a bunch of `string` / `Span` causing issues. 

Let me know when you find them. Pretty easy to fix.
